### DISCUSSION TITLE
3.0.0 Breaking Changes

### DISCUSSION THREAD

**Discussion by wizard-28** [pacstall/pacstall-programs#1599] [Announcements]
# Breaking changes

## What is happening

Pacstall will be releasing v3.0.0 Cobalt soon. We follow [Semantic Versioning](https://semver.org), and as such, the change from `2.0.1` -> `3.0.0` means that breaking changes have been introduced.

## How is this happening

https://github.com/pacstall/pacstall/pull/636 switches the package installation procedure from stowing to making native Debs, which is a long overdue change that we finally have gotten around to fixing.

### Pacscript developers

This means that packages containing symlinking commands such as `ln` now cannot be run inside `prepare`, `build`, or `install`. The only function that they should be used in is `postinst`, as the package contents will be unpacked before `posinst` is run, and they should be unlinked in `removescript`. Additionally, `-deb` pacscripts should include the `gives` variable, which should be assigned to the actual package name (IE: `vscodium-deb` should have `gives="codium"`). Also, you **must** make sure that your `depends` and `build_depends` are correct, and that packages are not mixed between them, because now `build_depends` will install first, and `depends` will install during the Deb installation, meaning if you currently have packages that should be in `build_depends`, but are in `depends`, they will not work.

### Pacstall users

If you update/reinstall your packages after upgrading to `3.0.0`, they will be reinstalled as a deb. You, as the end user should not see any major differences in the way you use Pacstall, but if you see anything strange or broken, please create an issue [here](https://github.com/pacstall/pacstall/issues/new?assignees=wizard-28&labels=bug%2Ctriage&template=BUG-REPORT.yml).

## What are we doing, and what should you do

Pacscript maintainers should test all their packages in a docker running the develop branch just to make sure their packages work, along with helping us find bugs. Pacstall devs will also be checking scripts at random during this time as well. We may create PRs that refactor your scripts to move around symlinking commands to places that will work on 3.0.0. They will continue to work on current versions as well.

## Additional information

Until release, the `develop` branch should be considered unstable, with bugfixes being introduced quickly. Please look [here](https://github.com/pacstall/pacstall/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+label%3Abash) for PRs that may fix issues you encounter, along with [this mega-PR](https://github.com/pacstall/pacstall/pull/643) that we are actively updating during this time.

When transitioning to symlinking in `postinst`, you must remember that the package has already been installed, meaning that you should not be using variables such as `STOWDIR` or `pkgdir`. Here is an example:

```bash
postinst() {
	sudo ln -s /opt/myprogram/bin/myprogram /bin/myprogram
}

removescript() {
	sudo rm -f /bin/myprogram
}
```

Additionally, because we are moving away from using Stow, the `STOWDIR` variable that is used in almost all scripts is a misnomer. You are free to continue using it in packages, or you can switch to the already existing `pkgdir` variable which is the full path to the installation directory (`$STOWDIR/$name`). It is not likely to be deprecated, as the variable has been around since the beginning of Pacstall.

CC: @pacstall/pacscript-maintainers

**Comment by Elsie19** [comment]
* `yabridge`
* `ardour-git`
* `postman-bin` (https://github.com/pacstall/pacstall-programs/pull/1605)
* `i3-gaps`
* `exploitdb-git`
* `geekbench-gin`
* `onnxruntime-1-7-0-bin`
* `smartgit`
* `gfn-electron-bin`
* `tidal-hifi-bin`
* `pycharm-community-bin`
* `android-studio`
* `awesome-git`
* `arduino-ide-bin`
* `wine-staging-git`

These are the packages that use symlinking commands

**Reply by echometerain** [reply]
the symlinks in exploitdb-git is alread in postinst()

**Reply by echometerain** [reply]
smartgit also

**Reply by Zahrun** [reply]
onnxruntime-1-7-0-bin, geekbench-bin, pycharm-community-bin, android-studio also
